draft-pda-core.md

```diff
@@ -291,11 +291,437 @@ Each PA task in this document is divided into three sub-protocols as follows.
+1. The client knows the URL of the leader endpoint, e.g., `example.com/metrics`.
+   We write this URL as `[leader]` below. (We write `[helper]` for a helper's
+   URL.)
+1. The client and leader can establish a leader-authenticated TLS channel.
+1. The leader and each helper can establish a leader-authenticated TLS channel.
+1. Each helper has chosen an HPKE key pair.
+1. The aggregators agree on a set of PA tasks, as well as the PA protocol and
+   parameters used for each task.
+
+## Upload {#pa-upload}
+
+[TODO: Add an illustration of this sub-protocol.]
+
+Uploading a report involves two requests to the leader. In the  *upload start
+request*, the client discovers the protocol-specific parameters and the endpoint
+URL of each helper. In the *upload finish request* uploads its report to the
+leader.
+
+[OPEN ISSUE: @acmiyaguchi pointed out that the use of an anonymizing proxy for
+uploading shares might be easier to implement if the "upload" phase involved a
+single HTTP request. Is this true? Does OHTTP
+(https//www.ietf.org/archive/id/draft-thomson-http-oblivious-01.html) allow
+clients to make multiple requests without too much fuss?]
+
+### Upload Start
+
+**Request.**
+Let `[leader]` denote the URL of the leader's report ingestion endpoint. The
+client sends a POST request to `[leader]/upload_start` with the following
+message:
+
+~~~
+struct {
+  PATask task;
+  opaque id[16];
+} PAUploadStartReq;
+~~~
+
+The first field, `task` corresponds to the PA task for which a report will be
+generated. The second, `id` is a random, 16-byte string generated by the
+client.
+
+[OPEN ISSUE: Maybe use a UUID for `PAUploadSTartReq.id`, as well as for
+`PATask.id`?]
+
+**Response.**
+The leader responds to well-formed requests to `[leader]/upload_start` with
+status 200 and the following message:
+
+~~~
+struct {
+  Url helper_urls<0..2^16-1>;
+  PAProto proto;
+  select (PAUploadStartResp.proto) {
+    case prio: PrioClientParam; // Section 3
+    case hits: HitsClientParam; // Section 4
+  }
+} PAUploadStartResp;
+~~~
+
+The message includes the URL of each helper and any protocol specific parameters
+the client needs to generate its report. The leader's response to malformed
+requests is specified in {{pa-error-common-aborts}}.
+
+### Upload Finish
+
+For each URL `[helper]` in `PAUploadStartResp.helper_urls`, the client sends a
+GET request to `[helper]/key_config`. The helper responds with status 200 and an
+`HPKEKeyConfig` message. Next, the client collects the set of helpers it will
+upload shares to. It ignores a helper if:
+* the client and helper failed to establish a secure, helper-authenticated
+  channel;
+* the GET request to the helper URL failed or didn't return a valid key config;
+  or
+* the key config specifies a KEM, KDF, or AEAD algorithm the client doesn't
+  recognize.
+
+If the set of supported helpers is empty, then the client aborts and alerts the
+leader with "no supported helpers". Otherwise, for each supported helper the
```

Comment:
Why is this alert important? What is the leader to do with this information?

Reply:
Can't the leader determine if helpers are "online"? This seems to reveal information that's specific to clients. Imagine, for example, that clients are prohibited from talking to helpers but not the leader. Is it OK that leaders learn that about a client? I'm not sure, so I'd be inclined to remove this unless we have a concrete use case.